src/client/java/net/neoforged/neoforge/client/neo3d/GpuDeviceProperties.java

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright (c) Forge Development LLC and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.client.neo3d;
+
+import com.mojang.blaze3d.platform.DepthTestFunction;
+import com.mojang.blaze3d.platform.DestFactor;
+import com.mojang.blaze3d.platform.SourceFactor;
+import com.mojang.blaze3d.shaders.ShaderType;
+import com.mojang.blaze3d.shaders.UniformType;
+import com.mojang.blaze3d.textures.AddressMode;
+import com.mojang.blaze3d.textures.TextureFormat;
+import java.util.Set;
+import org.jetbrains.annotations.ApiStatus;
+
+public interface GpuDeviceProperties {
```

Comment:
I think we should only introduce properties that matter in the context of what we are extending. Most properties here can be removed as they don't achieve the goal of this PR (making Blaze3D extensible).

Reply:
most of the ones added are covering existing B3D features, though they could be pulled from this PR and done in a second one instead if thats preferred.

Reply:
The difference is that features are enableable, where properties is reporting stuff about it and there may be none or many for each feature. Notable that this pattern was pulled from how VK handles it. Properties are just limit reporting, while features is telling it what things you want to use.

Reply:
not-enabled and unknown are two different issues IMO. The "unknown" catch is there for a"this _will_ explode in prod" catch, while using something not-enabled may still work and should probably be a different validation error.
As another option, some things maybe we don't want to use explicit enablement and instead could rely on what is reported for if its enabled or not. Depth test functions seems like a perfect example of that where there is no need to enable them explicitly, just check if the backend knows about the one you want to use.

Reply:
Fair enough, though maybe there should be second properties object that is created after device creation? Describing what is actually allowed to be used, rather than what the underlying backend is theoretically capable of. Properties as-is was mostly aimed at the latter case.

so, same interface, but the one you get from the created device will reflect that the extended depth tests weren't enabled, and therefore aren't allowed. The validation layer could also do strict enforcement on this,

Reply:
I think it makes sense to be able to query during the event whether the feature is available, and then later whether it was enabled. Same interface sounds fine. I'd maybe document clearly in the interface javadoc that it either represents the "potentially available" features or the enabled features depending on event usage vs. later usage?

Reply:
besides interface javadoc, the two would also be gotten from very different places, that can also be named to imply that. ie: from the event is "maximalProperties" vs from the device "enabledProperties" _will need to find actually good names for that_

Reply:
Most properties removed.

Comments reflect the intention of there being two different sets, however as there are no feature extensions included in this the same properties object (record) is still used in both circumstances.